diff -ru dosbox-staging-orig/include/pinhacks.h dosbox-staging/include/pinhacks.h
--- dosbox-staging-orig/include/pinhacks.h	2022-06-17 14:34:47.047090178 +0200
+++ dosbox-staging/include/pinhacks.h	2022-06-17 14:31:36.392623908 +0200
@@ -0,0 +1,25 @@
+/*
+ * Copyleft:
+ * This file was made to be used with DOSBOX, and so
+ * is under GNU general public license
+ */
+
+#ifndef __DOSBOX_PINHACK_H
+#define __DOSBOX_PINHACK_H
+#define PINHACKVERSION 3 
+
+extern struct scrollhack {
+	bool enabled,trigger;
+	struct { 
+		int min, max;
+	}triggerwidth,triggerheight;
+	struct {int height, width;} expand;
+	struct {struct { bool enabled,trigger;}
+		psychopinball,
+		pinballdreams;
+	}specifichack;
+	const char * doublewidth;
+}pinhack ;
+  
+#endif /* __DOSBOX_PINHACK_H */
+
diff -ru dosbox-staging-orig/src/dos/dos_files.cpp dosbox-staging/src/dos/dos_files.cpp
--- dosbox-staging-orig/src/dos/dos_files.cpp	2022-06-17 14:33:57.646451196 +0200
+++ dosbox-staging/src/dos/dos_files.cpp	2022-06-17 14:11:22.204916816 +0200
@@ -32,6 +32,7 @@
 #include "cross.h"
 #include "string_utils.h"
 #include "support.h"
+#include "pinhacks.h"
 
 #define DOS_FILESTART 4
 
@@ -538,6 +539,12 @@
 	if (flags>2) LOG(LOG_FILES,LOG_ERROR)("Special file open command %X file %s",flags,name);
 	else LOG(LOG_FILES,LOG_NORMAL)("file open command %X file %s",flags,name);
 
+        if ( pinhack.enabled && pinhack.specifichack.pinballdreams.enabled ) {
+                if ( !strcmp(name,"flippers.spr") ) { printf("PINHACK: pinhackpd: flippers.spr loaded, hack should trigger on next resolution change.\n");
+                pinhack.specifichack.pinballdreams.trigger=true; };
+        }
+
+
 	Bit16u attr = 0;
 	Bit8u devnum = DOS_FindDevice(name);
 	bool device = (devnum != DOS_DEVICES);
Seulement dans dosbox-staging/src/dos: dos_files.cpp.orig
diff -ru dosbox-staging-orig/src/dosbox.cpp dosbox-staging/src/dosbox.cpp
--- dosbox-staging-orig/src/dosbox.cpp	2022-06-17 14:33:57.650451248 +0200
+++ dosbox-staging/src/dosbox.cpp	2022-06-17 14:12:42.713956189 +0200
@@ -51,6 +51,7 @@
 #include "pci_bus.h"
 #include "midi.h"
 #include "hardware.h"
+#include "pinhacks.h"
 
 #if C_NE2000
 //#include "ne2000.h"
@@ -147,6 +148,19 @@
 void increaseticks();
 bool mono_cga=false;
 
+scrollhack pinhack;
+
+void parsetriggerrange (const char*&range,int &min, int &max, char delim) { // PINHACK stuff
+       char* range_delim = const_cast<char*>(strchr(range,delim));
+               if(range_delim && * range_delim) {
+                       min = atoi(range);
+                       max = atoi(range_delim+1);
+               }
+               else if (!range_delim) { // single value given
+                       max =  min = atoi(range);
+               }
+}
+
 static Bitu Normal_Loop()
 {
 	Bits ret;
@@ -404,6 +418,30 @@
 	else E_Exit("DOSBOX:Unknown machine type %s",mtype.c_str());
 }
 
+static void PINHACK_Init(Section * sec) {
+	Section_prop * section=static_cast<Section_prop *>(sec);
+	/* PINHACK config file parsing */
+	pinhack.enabled=(section->Get_bool("pinhack"));
+	const char* pinhacktriggerwidthrange=section->Get_string("pinhacktriggerwidth");
+	const char* pinhacktriggerheightrange=section->Get_string("pinhacktriggerheight");
+	pinhack.doublewidth=section->Get_string("pinhackdoublewidth");
+	pinhack.expand.height=(section->Get_int("pinhackexpandheight"));
+	pinhack.expand.width=(section->Get_int("pinhackexpandwidth"));
+	parsetriggerrange ( pinhacktriggerheightrange, pinhack.triggerheight.min, pinhack.triggerheight.max, '-' );
+	parsetriggerrange ( pinhacktriggerwidthrange, pinhack.triggerwidth.min, pinhack.triggerwidth.max, '-' );
+	if ( ( pinhack.triggerwidth.min == 0 ) && ( pinhack.triggerwidth.max == 0 ) ) pinhack.triggerwidth.max=9999;
+	printf("PINHACK: Your DOSBox has been hacked with pinball hacks V %d ",PINHACKVERSION);
+	if (pinhack.enabled) { 
+		printf("and enabled in your config file!\n");
+		printf("PINHACK: width settings: trigger at:");
+		printf("%d-%d, expand to %d\n",pinhack.triggerwidth.min, pinhack.triggerwidth.max,pinhack.expand.width);
+		printf("PINHACK: height settings: trigger at:");
+		printf("%d-%d, expand to %d\n",pinhack.triggerheight.min, pinhack.triggerheight.max,pinhack.expand.height);
+		printf("PINHACK: doublewidth: %s\n",pinhack.doublewidth);
+	}
+	else printf("but disabled in your config file!\n");
+	pinhack.specifichack.pinballdreams.enabled=(section->Get_bool("pinhackpd"));
+}
 
 void DOSBOX_Init(void) {
 	Section_prop * secprop;
@@ -433,6 +471,42 @@
 		"hercules", "cga", "cga_mono", "tandy", "pcjr", "ega",
 		"vgaonly", "svga_s3", "svga_et3000", "svga_et4000",
 		"svga_paradise", "vesa_nolfb", "vesa_oldvbe", 0 };
+		
+	// PINHACK: pinhack config file section
+
+	secprop=control->AddSection_prop("pinhack",&PINHACK_Init, true);
+
+	Pbool = secprop->Add_bool("pinhack",Property::Changeable::Always,false);
+	Pbool->Set_help("Boolean: Enable pinball hacks to display whole table at once. Not enabled per default.");
+
+	Pstring = secprop->Add_string("pinhacktriggerwidth",Property::Changeable::Always,"0");
+	Pstring->Set_help("The X resolution (width) the pinball hack should trigger at. It is not checked by default or if set to 0. Can be a range.");
+
+	Pstring = secprop->Add_string("pinhacktriggerheight",Property::Changeable::Always,"350");
+	Pstring->Set_help("The Y resolution (height) the pinball hack should trigger at. Default is 350 (good for Pinball Fantasies). Can be a range.");
+
+	Pint = secprop->Add_int("pinhackexpandwidth",Property::Changeable::Always,0);
+	Pint->SetMinMax(0,4000);
+	Pint->Set_help("The X resolution (width) DOSBox will expand to if pinball hack is enabled and triggers.\n"
+			"Not very useful probably, but provided here in case someone finds a game where it is useful!");
+
+	Pint = secprop->Add_int("pinhackexpandheight",Property::Changeable::Always,608);
+	Pint->SetMinMax(1,4000);
+	Pint->Set_help("The Y resolution (height) DOSBox will expand to if pinball hack is enabled and triggers.");
+
+	Pbool = secprop->Add_bool("pinhackpd",Property::Changeable::Always,false);
+	Pbool->Set_help("Boolean: Enable pinball dreams to work correctly during intros. Default:false\n");
+
+	Pbool = secprop->Add_bool("pinhackpsycho",Property::Changeable::Always,false);
+	Pbool->Set_help("Boolean: Psycho Pinball tilt craphics are messed up if hack enabled. TODO, this is a placeholder and has no function currently.");
+
+	Pstring = secprop->Add_string("pinhackdoublewidth",Property::Changeable::Always,"normal");
+	Pstring->Set_help("Used to enable forcing (or disabling) doublewidth with pinhack. The original patch disabled it,\n"
+			"but you may find it better enabled, or left to decide by dosbox and use aspect ratio correction insread.\n"
+			"normal=do not touch the setting, let DOSBox decide. yes=doublewidth. no=no doublewidth");
+
+// PINHACK: end config file section
+		
 	secprop=control->AddSection_prop("dosbox",&DOSBOX_RealInit);
 	Pstring = secprop->Add_path("language",Property::Changeable::Always,"");
 	Pstring->Set_help("Select another language file.");
Seulement dans dosbox-staging/src: dosbox.cpp.orig
diff -ru dosbox-staging-orig/src/hardware/vga_crtc.cpp dosbox-staging/src/hardware/vga_crtc.cpp
--- dosbox-staging-orig/src/hardware/vga_crtc.cpp	2022-06-17 14:33:57.658451351 +0200
+++ dosbox-staging/src/hardware/vga_crtc.cpp	2022-06-17 14:10:50.104502379 +0200
@@ -25,6 +25,7 @@
 #include "cpu.h"
 #include "video.h"
 #include "pic.h"
+#include "pinhacks.h"
 
 #define crtc(blah) vga.crtc.blah
 
@@ -190,11 +191,13 @@
 		*/
 		break;
 	case 0x0C:	/* Start Address High Register */
+if (pinhack.trigger) val=0;
 		crtc(start_address_high)=val;
 		vga.config.display_start=(vga.config.display_start & 0xFF00FF)| (val << 8);
 		/* 0-7  Upper 8 bits of the start address of the display buffer */
 		break;
 	case 0x0D:	/* Start Address Low Register */
+if (pinhack.trigger) val=0;
 		crtc(start_address_low)=val;
 		vga.config.display_start=(vga.config.display_start & 0xFFFF00)| val;
 		/*	0-7	Lower 8 bits of the start address of the display buffer */
diff -ru dosbox-staging-orig/src/hardware/vga_draw.cpp dosbox-staging/src/hardware/vga_draw.cpp
--- dosbox-staging-orig/src/hardware/vga_draw.cpp	2022-06-17 14:33:57.662451403 +0200
+++ dosbox-staging/src/hardware/vga_draw.cpp	2022-06-17 14:10:50.104502379 +0200
@@ -25,6 +25,8 @@
 #include "../gui/render_scalers.h"
 #include "vga.h"
 #include "pic.h"
+#include "pinhacks.h"
+#include <cstdio>
 
 //#undef C_DEBUG
 //#define C_DEBUG 1
@@ -725,6 +727,7 @@
 			vga.draw.address+=vga.draw.address_add;
 		}
 		vga.draw.lines_done++;
+       if (pinhack.trigger == false) {
 		if (vga.draw.split_line==vga.draw.lines_done) {
 #ifdef VGA_KEEP_CHANGES
 			VGA_ChangesEnd( );
@@ -733,6 +736,7 @@
 #ifdef VGA_KEEP_CHANGES
 			vga.changes.start = vga.draw.address >> VGA_CHANGE_SHIFT;
 #endif
+               }
 		}
 	}
 	if (--vga.draw.parts_left) {
@@ -1546,6 +1550,25 @@
 			height/=2;
 		}
 	}
+	if (pinhack.enabled) {    // Enabled in config?
+	printf("PINHACK: "); // Check for tirggering preconditions...
+	if ( ( !pinhack.specifichack.pinballdreams.enabled || pinhack.specifichack.pinballdreams.trigger ) && 
+            ( height>=pinhack.triggerheight.min && height<=pinhack.triggerheight.max ) &&
+            ( pinhack.triggerwidth.min<=width && width<=pinhack.triggerwidth.max ) ) {
+              	printf("triggered -> ");
+              	pinhack.trigger = true;
+           	printf("original geometry: %dx%d. expanding to geometry: %dx%d.\n",width,height,
+			pinhack.expand.width ? pinhack.expand.width : width,
+			pinhack.expand.height ? pinhack.expand.height : height);
+		// If width or height not set, do not expand!
+		if ( pinhack.expand.height ) height = pinhack.expand.height;
+		if ( pinhack.expand.width ) width = pinhack.expand.width;
+              	if ( pinhack.specifichack.pinballdreams.trigger ) pinhack.specifichack.pinballdreams.trigger=false; // On next resolution change, return to normal
+       } else {
+              	pinhack.trigger = false;
+              	printf("trigger values evalueted but not triggered! Current resolution: %dx%d\n",width,height);
+       };
+}
 	vga.draw.lines_total=height;
 	vga.draw.parts_lines=vga.draw.lines_total/vga.draw.parts_total;
 	vga.draw.line_length = width * ((bpp + 1) / 8);
@@ -1598,6 +1621,11 @@
 
 		vga.draw.width = width;
 		vga.draw.height = height;
+       if ( pinhack.trigger ) {
+               if ( pinhack.doublewidth == "yes" ) doublewidth=true;
+               if ( pinhack.doublewidth == "no" ) doublewidth=false;
+               printf("PINHACK: Doublewidth: %s\n",pinhack.doublewidth);
+       }
 		vga.draw.doublewidth = doublewidth;
 		vga.draw.doubleheight = doubleheight;
 		vga.draw.aspect_ratio = aspect_ratio;
